In [9]:
# uncomment this command if you get an error from the following cell saying that 'highs' solver is not recognized
# !pip install --upgrade --user scipy


In [10]:
import numpy as np
from scipy.optimize import linprog

# Define cost per serving (Chicken, Rice, Broccoli)
cost = [2.50, 0.80, 1.50]  # Minimize: 2.5x + 0.8y + 1.5z

# Define nutrition constraints matrix (Calories, Protein, Fat, Vitamin C)
nutrition = [
    [250, 130, 50],    # Calories per serving
    [30, 3, 4],        # Protein per serving
    [8, 0.5, 0.2],     # Fat per serving
    [0, 0, 50]         # Vitamin C per serving
]

# Define minimum daily intake for each nutrient
min_nutrition = [2000, 50, 10, 30]  # [Calories, Protein, Fat, Vitamin C]

# Convert constraints to standard form (Ax ≤ b)
A_ub = -1 * np.array(nutrition)  # Flip sign to convert ≥ to ≤
b_ub = -1 * np.array(min_nutrition)

# Define bounds for each food item (no negative servings)
bounds = [(0, None), (0, None), (0, None)]

# Solve the linear programming problem
result = linprog(c=cost, A_ub=A_ub, b_ub=b_ub, bounds=bounds, method="highs")

# Print results
if result.success:
    servings = result.x
    print("Optimal daily servings:")
    print(f"  Chicken: {servings[0]:.2f} servings")
    print(f"  Rice: {servings[1]:.2f} servings")
    print(f"  Broccoli: {servings[2]:.2f} servings")
    print(f"Total cost: ${result.fun:.2f}")
else:
    print("Optimization failed:", result.message)



Optimal daily servings:
  Chicken: 0.33 servings
  Rice: 14.52 servings
  Broccoli: 0.60 servings
Total cost: $13.34


In [11]:
#____________________________________________________Answers____________________________________________________

1. Before adding Vitamin C: What was the optimal diet and cost?

The optimal servings per day were:

Chicken: 0.33 servings

Rice: 14.75 servings

Broccoli: 0.00 servings

The minimum cost of the diet was $12.62 per day


_______________________________________________________________________________

2. After adding Vitamin C: How did the results change? Did the cost increase?

The new optimal servings per day were:

Chicken: 0.33 servings

Rice: 14.52 servings

Broccoli: 0.60 servings

The new minimum cost was $13.34 per day

Changes observed:
The amount of Broccoli likely increased to meet the Vitamin C requirement.

The cost increased because adding the Vitamin C constraint forced the diet to include more Broccoli, which is relatively more expensive than Rice.

Before adding the Vitamin C constraint, the optimization minimized cost by choosing a combination of Chicken, Rice, and Broccoli that met the basic calorie, protein, and fat requirements at the lowest price.

However, after adding the requirement of at least 30 mg of Vitamin C per day, the solver had to ensure that enough Broccoli was included in the diet. Since Broccoli costs $1.50 per serving, this led to a higher total cost compared to the previous solution, where the optimizer might have relied more on cheaper options like Rice.

_______________________________________________________________________________

3. What does this tell you about trade-offs in diet planning?

Adding new constraints (such as Vitamin C intake) forces changes in food selection.

Broccoli consumption increased, affecting the cost and balance of other nutrients.

This demonstrates a key trade-off:

Health vs. Cost – A healthier diet may be more expensive.

Flexibility vs. Constraints – More dietary requirements reduce flexibility in choosing cheaper foods.

In real-world diet planning, nutritional balance and affordability must be optimized together.

_______________________________________________________________________________